File: ios/YATA/Infrastructure/User/UserMemoryRepository.swift
//
//  UserMemoryRepository.swift
//  YATA
//
//  Created by Alpár Szotyori on 26.05.24.
//

import Foundation

class UserMemoryRepository: UserRepository {

    private var users: [User] = []

    func findActiveUser() -> Result<YATA.User?, any Error> {
        Result.success(users.first { user in user.isActive})
    }

    func getAllUsers() -> Result<[YATA.User], any Error> {
        Result.success(users)
    }

    func updateUsers(_ users: [YATA.User]) -> Result<Void, any Error> {
        users.forEach { user in
            _ = updateUser(user)
        }
        return Result.success(())
    }

    func addUser(_ user: YATA.User) -> Result<Void, any Error> {
        users.append(user)
        return Result.success(())
    }

    func updateUser(_ user: YATA.User) -> Result<Void, any Error> {
        if let index = users.firstIndex(where: { $0.id == user.id }) {
            users[index] = user
        }
        return Result.success(())
    }
}

File: android/app/src/main/java/de/repeatuntil/yata/infrastructure/user/UserMemoryRepository.kt
package de.repeatuntil.yata.infrastructure.user

import de.repeatuntil.yata.domain.user.User
import de.repeatuntil.yata.domain.user.UserRepository

class UserMemoryRepository : UserRepository {

    private val users: MutableList<User> = mutableListOf()

    override suspend fun findActiveUser(): Result<User?> {
        return Result.success(users.firstOrNull { it.isActive })
    }

    override suspend fun getAllUsers(): Result<List<User>> {
        return Result.success(users)
    }

    override suspend fun updateUsers(users: List<User>): Result<Unit> {
        users.forEach { updateUser(it) }
        return Result.success(Unit)
    }

    override suspend fun addUser(user: User): Result<Unit> {
        users.add(user)
        return Result.success(Unit)
    }

    override suspend fun updateUser(user: User): Result<Unit> {
        val index = users.indexOfFirst { it.id == user.id }
        if (index >= 0) {
            users[index] = user
        }
        return Result.success(Unit)
    }
}